proof/refine/ARM/Ipc_R.thy

```diff
@@ -2082,60 +2102,30 @@ lemma replyPush_corres:
-        apply (clarsimp simp: tcb_relation_def)
-       apply (wpsimp wp: thread_get_wp')
+              apply (wpsimp wp: hoare_vcg_imp_lift' hoare_vcg_all_lift)
+             apply (clarsimp simp: valid_tcb_state'_def cong: conj_cong)
+             apply (wpsimp wp: hoare_vcg_imp_lift' hoare_vcg_all_lift updateReply_valid_objs')
+            apply (clarsimp simp: tcb_relation_def)
+           apply (wpsimp wp: thread_get_wp, assumption)
+          apply (wpsimp wp: threadGet_wp)
+         apply (clarsimp simp: tcb_relation_def)
+        apply (wpsimp wp: thread_get_wp')
       apply (wpsimp wp: threadGet_wp)
      apply (subgoal_tac "caller \<noteq> reply_ptr")
       apply (subgoal_tac "caller \<noteq> idle_thread_ptr")
-       apply (clarsimp simp: invs_def valid_state_def
-                             valid_pspace_def cong: conj_cong)
+       apply (clarsimp simp: cong: conj_cong)
+       apply (subst tcb_obj_at_equiv[where g=callee], clarsimp simp: obj_at_def, clarsimp simp: is_tcb)
        apply (frule valid_objs_valid_tcbs, clarsimp)
        apply (frule (1) valid_objs_ko_at[where ptr=caller])
-       apply (subgoal_tac "ex_nonz_cap_to caller s", clarsimp)
-        apply (frule (1) idle_no_ex_cap)
-        apply (frule (2) no_tcb_not_in_replies_with_sc, clarsimp)
-        apply (intro conjI)
             apply (clarsimp simp: valid_obj_def valid_tcb_def)
-           apply clarsimp
-          apply (clarsimp simp: replies_blocked_def pred_tcb_at_def obj_at_def)
-         apply (erule delta_sym_refs_insert_only, simp)
-           apply (subst set_eq_subset, intro conjI)
-            apply (clarsimp simp: state_refs_of_def)
-           apply (clarsimp simp: state_refs_of_def obj_at_def is_tcb get_refs_def2)
-           apply (subgoal_tac "tcb_st_refs_of (tcb_state tcb) = {}", simp)
-            apply fastforce
-           apply (fastforce simp: tcb_st_refs_of_def pred_tcb_at_def obj_at_def)
-          apply (subst set_eq_subset, intro conjI)
-           apply (clarsimp simp: state_refs_of_def)
-          apply (clarsimp simp: state_refs_of_def obj_at_def is_reply get_refs_def2)
-          apply (clarsimp simp: obj_at_def sk_obj_at_pred_def)
-         apply simp
-        apply (subgoal_tac "sc_tcb_sc_at ((=) (Some caller)) y s")
-         apply (clarsimp simp: sc_at_pred_n_def)
-         apply (erule (1) if_live_then_nonz_capD)
-         apply (clarsimp simp: live_def live_sc_def, fastforce)
-        apply (subst sym_refs_bound_sc_tcb_iff_sc_tcb_sc_at[symmetric, OF refl eq_commute])
-         apply assumption
-        apply (clarsimp simp: pred_tcb_at_def obj_at_def)
-       apply (erule (1) if_live_then_nonz_capD)
-       apply (clarsimp simp: live_def  is_tcb obj_at_def)
-       apply (frule (1) bound_sc_tcb_at_idle_sc_idle_thread[where t=caller])
-        apply (clarsimp simp: pred_tcb_at_def obj_at_def)
-       apply simp
-      apply clarsimp
-      apply (frule invs_valid_idle)
+           apply (subst sc_at_ppred_exm; clarsimp)
```

Comment:
I would have thought this would solve at most one subgoal, so maybe indentation is off